### PR TITLE
Standalone App Non-Home Index Fixes

### PR DESCRIPTION
**Pull request by ckrew**
This PR looks to resolve https://github.com/tethysplatform/tethys/issues/1042. The issue is that there are many redirects and reverses that point to a hardcoded "home" url. While this works fine when the single app index is home, these redirects fail when an app index is not "home". 

The PR resolves the issue by putting conditional logic in front of the hardcoded "home" redirects. If the tethys portal is in single app mode, then the "home" redirects will point towards the app index url instead. 

Along with the hardcoded "home" fix, I have also updated additional redirects that point to the app_library url which is no longer available in standalone app mode. If the tethys portal is in single app mode, these old app_library redirects will point to more appropriate endpoints instead. 

This PR also adds a warning to users on all pages when single app mode is enabled but there are no applications installed.

image

### CODE REVIEW

**Review by swainn — APPROVED** [review]
Approve once tests are passing

**Comment by ckrew** [comment]
This PR also adds a warning to users on all pages when single app mode is enabled but there are no applications installed.

image

File: tethys_apps/utilities.py [hunk 1]
```diff
@@ -620,13 +621,18 @@ def get_configured_standalone_app():
...
-        app = TethysApp.objects.get(package=standalone_app)
-    else:
-        app = TethysApp.objects.first()
-        if not app:
-            raise ObjectDoesNotExist("No Tethys Apps have been installed")
+    try:
+        if standalone_app:
+            app = TethysApp.objects.get(package=standalone_app)
+        else:
+            app = TethysApp.objects.first()
```
Comment (swainn) [inline tethys_apps/utilities.py:630]: What happens if the database is configured, but there is no app installed? Do these raise an exception? We need to catch those exceptions too.

File: tethys_apps/utilities.py [hunk 2]
```diff
@@ -620,13 +621,18 @@ def get_configured_standalone_app():
...
+    except ProgrammingError:
+        # When MULTIPLE_APP_MODE is enabled but there is no app installed, then setting up the tethys DB fails because
+        # the code is trying to fetch data from a table that doesn't exist. Catch this exception and pass so setup can
```
Comment (swainn) [inline tethys_apps/utilities.py:633]:
```suggestion
        # ProgrammingError: the code is trying to fetch data from a table that doesn't exist. Catch this exception and pass so setup can
```

File: tethys_portal/context_processors.py [hunk 3]
```diff
@@ -30,15 +31,22 @@ def tethys_portal_context(request):
...
+def check_single_app_mode(request):
     if settings.MULTIPLE_APP_MODE:
         return False, None
     else:
-        return True, get_configured_standalone_app().name
+        configured_app = get_configured_standalone_app()
+        if not configured_app:
+            messages.warning(
+                request,
+                "MULTIPLE_APP_MODE is disabled but there is no configured tethys application.",
```
Comment (swainn) [inline tethys_portal/context_processors.py:49]:
```suggestion
                "MULTIPLE_APP_MODE is disabled but there is no Tethys application installed.",
```

File: tethys_apps/utilities.py [hunk 4]
```diff
@@ -620,13 +621,18 @@ def get_configured_standalone_app():
...
-    else:
-        app = TethysApp.objects.first()
-        if not app:
-            raise ObjectDoesNotExist("No Tethys Apps have been installed")
+    try:
+        if standalone_app:
+            app = TethysApp.objects.get(package=standalone_app)
+        else:
+            app = TethysApp.objects.first()
+    except ProgrammingError:
```
Comment (sdc50) [inline tethys_apps/utilities.py:631]:
```suggestion
    except (ProgrammingError, TethysApp.DoesNotExist):
```

File: tethys_portal/templates/header.html [hunk 5]
```diff
@@ -27,11 +27,21 @@
...
+                {% url 'user:profile' as header_url  %} 
+              {% endif %}
+            {% else %}
+              {% url 'app_library' as header_url %}
+            {% endif %}
             {% if ENABLE_OPEN_PORTAL or user.is_authenticated and user.is_active %}
-              <li class="nav-item">
-                <a class="h5 me-3 mb-0 nav-link{% if request.path == app_library_url %} active" aria-current="page{% endif %}" href="{{ app_library_url }}" title="{% if single_app_name %}{{ single_app_name }}{% elif site_globals.apps_library_title %}{{ site_globals.apps_library_title }}{% else %}Apps{% endif %}">{% if single_app_name %}{{ single_app_name }}{% elif site_globals.apps_library_title %}{{ site_globals.apps_library_title }}{% else %}Apps{% endif %}</a>
-              </li>
+              {% if not single_app_mode or single_app_mode and configured_single_app.name %}
```
Comment (swainn) [inline tethys_portal/templates/header.html:40]: Move this logic to the context processor and use simple context variables like `app_library_name` and `app_library_url`. Set the values in the context processor.

File: tethys_portal/templates/header.html [hunk 6]
```diff
@@ -5,7 +5,7 @@
   <nav id="header-navbar" class="navbar navbar-expand-md navbar-dark shadow px-4 py-3 mx-3" style="background-color: {{ site_globals.primary_color|default:'#0a62a9' }};">
     <div class="container-fluid">
       {% block header_brand %}
-        <a id="header-brand" class="navbar-brand" href="{% url 'home' %}">
+        <a id="header-brand" class="navbar-brand" href="{% if single_app_mode %} {% if configured_single_app %} {% url configured_single_app.index_url %}  {% else %} {% url 'user:profile' %} {% endif %} {% else %} {% url 'home' %} {% endif %}">
```
Comment (swainn) [inline tethys_portal/templates/header.html:8]: Change this to use the App Name and App icon and don't show the app library link in single app mode.

**Review by swainn — COMMENTED** [review]
Add 15pm margin-right on the profile-admin do-hickey in the app template.

File: tests/unit_tests/test_tethys_portal/test_context_processors.py [hunk 7]
```diff
@@ -1,24 +1,59 @@
...
     def setUp(self):
         pass
 
     def tearDown(self):
         pass
 
-    @override_settings(MULTIPLE_APP_MODE=False)
-    def test_check_single_app_mode_single(self):
-        single_app_mode, single_app_name = context_processors.check_single_app_mode()
+    def test_context_processors_multiple_app_mode(self):
```
Comment (swainn) [inline tests/unit_tests/test_tethys_portal/test_context_processors.py:13]: Avoid assuming the values of settings. Use the settings.override here to set `MULTIPLE_APP_MODE` to `True` explicitly.

**Review by swainn — COMMENTED** [review]
One last comment. Check your tests to make sure you aren't assuming a certain setting for `MULTIPLE_APP_MODE` like what the default is. Always set settings explicitly for tests that depend on settings.

File: tests/unit_tests/test_tethys_portal/test_context_processors.py (hunk 7 above)
Comment (ckrew) [reply tests/unit_tests/test_tethys_portal/test_context_processors.py:13]: sounds good. Ill go through and update those tests

**Review by swainn — APPROVED** [review]

**Review by sdc50 — APPROVED** [review]

**Pull request merged**